BXMSDOC-5392-master-kogito: Create kogito configuration doc.

sterobin opened the pull request:
@cristianonicolai , @tiagodolphine , @ricardozanini , can you SME review this new "Configuring Kogito supporting services and runtime capabilities" doc for dev preview? Doc previews linked below.

**I had to guess on a lot of stuff with the data index and jobs services, so your careful verification would be very helpful.**

Cristiano or whoever knows about the Prometheus add-on, please search the several questions tagged **@Cristiano** in the output.

Note that some of the content is reused from the "Developing process services" doc, such as process instance management.

And I'm open to ideas for a better title.

Links:
* [JIRA](https://issues.redhat.com/browse/BXMSDOC-5372)
* [Enterprise doc preview](http://file.rdu.redhat.com/~sterobin/BXMSDOC-5392/) - **Refresh to get latest.**
* [Community doc preview](http://file.rdu.redhat.com/~sterobin/BXMSDOC-5392_COMM/#chap_kogito-configuring) - Same, plus "Data Index security" section excluded from enterprise, per Ricardo's suggestion. **Refresh to get latest.**

ricardozanini commented on doc-content/kogito-docs/src/main/asciidoc/configuration/con_kogito-supporting-services-and-configuration.adoc:
```diff
@@ -0,0 +1,24 @@
+
+* Infinispan persistence
+* Apache Kafka reactive messaging
+* Keycloack security
```
I'd remove any Keycloak/security topics for now since we won't support those formally.

sterobin replied:
Done.

ricardozanini commented on doc-content/kogito-docs/src/main/asciidoc/configuration/proc_data-index-service-security.adoc:
```diff
@@ -0,0 +1,59 @@
+[id='proc_data-index-service-security_{context}']
+
+= Enabling {PRODUCT} Data Index Service security with OpenID Connect
```
@cristianonicolai I'd say that we should not include this as base of our doc for DP. It's ok to have it in community docs, though. @sterobin is it possible to add conditionals on the "security" topic?

sterobin replied:
@ricardozanini , yes, it's already conditioned. Note that it's not in the enterprise rendered doc preview but is in the community rendered preview.

cristianonicolai replied:
+1

ricardozanini commented on doc-content/kogito-docs/src/main/asciidoc/configuration/proc_data-index-service-using.adoc:
```diff
@@ -0,0 +1,91 @@
+.Running the Data Index Service
+[source,subs="+quotes"]
+----
+$ java -jar data-index-service-__VERSION__-runner.jar
```
We are missing the authentication properties for Infinispan connection: 
1. First the user needs to grab Infinispan credentials from the server: https://infinispan.org/docs/stable/titles/server/server.html#security
2. Pass those credentials to Data Index:
```
java -Dquarkus.infinispan-client.auth-username=<user> \
-Dquarkus.infinispan-client.auth-password=<password> \
-jar data-index-service-__VERSION__-runner.jar
```

More info that can be added linking this article: https://quarkus.io/guides/infinispan-client#authentication

It's not in this article, but we can inform the `hotrod.properties` in the jar command using this prefix: `quarkus.infinispan-client`

sterobin replied:
@ricardozanini , updated this section and the runtime properties section with your feedback. Good now? **Refresh the doc previews**
* Updated run section: http://file.rdu.redhat.com/~sterobin/BXMSDOC-5392/#proc_data-index-service-using_kogito-configuring
* Updated runtime properties section: http://file.rdu.redhat.com/~sterobin/BXMSDOC-5392/#ref_kogito-runtime-properties_kogito-configuring

However, I don't understand what you're asking me to do with the `hotrod.properties`, nor do I see that exact property anywhere as I search online. Pinged you in chat.

ricardozanini replied:
Great work, @sterobin looks good to me. I'll left the final review to Cristiano and Tiago.

ricardozanini commented on doc-content/kogito-docs/src/main/asciidoc/configuration/ref_kogito-runtime-properties.adoc:
```diff
@@ -0,0 +1,100 @@
+
+Example: `kogito.messaging.as-cloudevents=true`
+
+.2+|Infinispan persistence
```
It's important to also include authentication options here.

sterobin replied:
Done

tiagodolphine commented on doc-content/kogito-docs/src/main/asciidoc/configuration/con_jobs-service.adoc:
```diff
@@ -0,0 +1,236 @@
+|Defines the maximum interval in milliseconds when retrying to execute jobs, in case the execution fails
+|Long type
+|`60000`
+|===
```
- Add Kafka properties:

|mp.messaging.outgoing.kogito-job-service-job-status-events.bootstrap.servers
|Kafka bootstrap servers address with the port used to publish events
|string
|localhost:9092

|mp.messaging.outgoing.kogito-job-service-job-status-events.topic
|Topic name where the events will be published
|string
|kogito-jobs-events

sterobin replied:
Done

tiagodolphine commented on doc-content/kogito-docs/src/main/asciidoc/configuration/con_jobs-service.adoc:
```diff
@@ -0,0 +1,236 @@
+To enable Infinispan persistence in the Jobs Service, you specify the property `-Dkogito.job-service.persistence=infinispan` during Jobs Service start-up or add the property to the `src/main/resources/application.properties` file in the Jobs Service project.
++
+For more information about Infinispan configuration with Quarkus applications, see https://quarkus.io/guides/infinispan-client[Infinispan client] in the Quarkus documentation.
+
```
- Add the Messaging section:

**Messaging**
Jobs Service publish events for each job state transition to a topic, in this way any application could subscribe to this topic to receive information of jobs and the state it was transitioned. Inside Kogito architecture, Data-Index service is subscribed to the jobs topic, indexing the jobs with their current, in this way the all jobs can be fetched directly in the Data-Index.

**Apache Kafka**
The current supported message broker is Apache Kafka.

**Apache Kafka Configuration**
To enable Kafka messaging, you should use a specific Quarkus profile events-support that could be set either using -D option during the startup -Dquarkus.profile=events-support or an environment variable QUARKUS_PROFILE=events-support. The Kafka bootstrap server can be set on the default Quarkus standard -Dmp.messaging.outgoing .kogito-job-service-job-status-events.bootstrap.servers={server address}. Other configurations like the topic name, can be set in the same way according to the _Configuration properties._

sterobin replied:
Done

tiagodolphine commented on doc-content/kogito-docs/src/main/asciidoc/configuration/proc_jobs-service-using.adoc:
```diff
@@ -0,0 +1,88 @@
+.Running the Jobs Service
+[source,subs="+quotes"]
+----
+$ java -jar data-index-service-__VERSION__-runner.jar
```
**data-index-service** it should be jobs-service.

sterobin replied:
Oops! Copy-pasted flub. Good catch! Fixed.

sterobin commented:
@tiagodolphine , implemented your feedback and updated the doc previews linked above. Let me know what you think. **Refresh the doc preview page to get latest.**

cristianonicolai commented on doc-content/kogito-docs/src/main/asciidoc/configuration/ref_data-index-service-queries.adoc:
```diff
@@ -0,0 +1,420 @@
+** `to`: Date time
+--
+
+Sort query results using `orderBy` parameter::
```
I commented a few of these, but it might actually needed.

sterobin commented on doc-content/kogito-docs/src/main/asciidoc/configuration/con_kogito-runtime-events.adoc:
```diff
@@ -0,0 +1,206 @@
+
+These events provide a complete view of the process instances being executed and can be consumed by an event listener, such as a `ProcessEventListener` configuration.
+
+IMPORTANT: Events are produced only when assets are executed successfully, without any errors during the course of execution.
```
@cristianonicolai , We've been discussing terms a bit for the glossary. The term "business assets" is just the generic term for processes, tasks, decisions, whatever that we've used since BA 7.x. In Kogito, BPMN models, DMN models, DRL units, XLS decision tables, etc. as a whole are apparently called "business models" (per team lead decision), but the individual components in each of these (processes, tasks, scripts, rules) I think we're still calling "assets" until we have a better alternative. Open to suggestions. For now, I've actually moved the note above to the parent section, and reworded a bit to this:

> NOTE: Events are produced only when assets such process instances or task instances are executed successfully, without any errors during the course of execution.

Wdyt?

cristianonicolai replied:
I think what I'm struggling here is that the way events work is that they are sent when the _unit of work_ runs without error. That is different from saying that the process instance had no error during the execution. So really the unit of work is a series of nodes that were executed in the process instance, like a transaction until it reaches either the end of the process of a wait state. On the technical side, we reference these as node instances. If we want to abstract further, from the BPMN specification point of view, I'm not sure there is something that matches to that but perhaps, Activities ( which are only tasks ) could be used here. Anyway, this is not a blocker to me.

sterobin commented on doc-content/kogito-docs/src/main/asciidoc/configuration/con_kogito-runtime-events.adoc:
```diff
@@ -0,0 +1,206 @@
+
+= {PRODUCT} runtime events
+
+A runtime event is record of a significant change of state in the application domain at a point in time. {PRODUCT} emits runtime events as a result of executed assets, such as a process instance or a task instance in a process, and can then use these events to notify third parties about changes to the BPMN model instance and its data.
```
Ha, saw that too when doing the glossary and did the same. Thanks.

sterobin commented on doc-content/kogito-docs/src/main/asciidoc/chap_kogito-configuring.adoc:
```diff
@@ -0,0 +1,24 @@
+:context: kogito-configuring
+
+// Purpose statement for the assembly
+As a developer of business processes and rules, you can configure {PRODUCT} supporting services and runtime properties for advanced use cases with your {PRODUCT} services.
```
Yep, you're right. Done. Updated in all docs throughout (developer of business processes and decisions, etc.) In some cases, we still distinguish between decisions and rules for clarity (like with DRL vs. DMN), but for the overall references like this, the general term should indeed be decisions. Thanks.

sterobin commented on doc-content/kogito-docs/src/main/asciidoc/configuration/proc_prometheus-metrics-monitoring.adoc:
```diff
@@ -0,0 +1,121 @@
+----
+scrape_configs:
+  job_name: 'kogito'
+metrics_path: /rest/metrics
```
Done

sterobin commented on doc-content/kogito-docs/src/main/asciidoc/configuration/proc_prometheus-metrics-monitoring.adoc:
```diff
@@ -0,0 +1,121 @@
+. To verify the metrics configuration, use a REST client or curl utility to send API requests with the following components:
++
+--
+* *URL*: `\http://localhost:8080/rest/metrics`
```
Done

sterobin commented on doc-content/kogito-docs/src/main/asciidoc/configuration/proc_prometheus-metrics-monitoring.adoc:
```diff
@@ -0,0 +1,121 @@
+* *HTTP headers*:
+** `accept`: `application/json`
+** `content-type`: `application/json`
+* *HTTP methods*: `GET`, `POST`, or `DELETE`
```
Ah, got it.  Thanks. So I've just reworded to this:

> To verify the metrics configuration, use a REST client or curl utility to send a `GET` request to the endpoint `\http://localhost:8080/metrics`.

sterobin commented on doc-content/kogito-docs/src/main/asciidoc/configuration/proc_prometheus-metrics-monitoring.adoc:
```diff
@@ -0,0 +1,121 @@
+.Example curl command to return Prometheus metrics
+[source]
+----
+curl -X GET http://localhost:8080/rest/metrics -H 'content-type: application/json' -H 'accept: application/json'
```
Done

sterobin commented on doc-content/kogito-docs/src/main/asciidoc/configuration/proc_prometheus-metrics-monitoring.adoc:
```diff
@@ -0,0 +1,121 @@
+.Example response (JSON)
+[source,json]
+----
+
```
Thanks!

sterobin commented on doc-content/kogito-docs/src/main/asciidoc/configuration/proc_prometheus-metrics-monitoring.adoc:
```diff
@@ -0,0 +1,121 @@
+.Example scrape configurations in `prometheus.yaml` file
+[source,yaml,subs="+quotes"]
+----
+scrape_configs:
```
Ah, cool. Okay.

sterobin commented on doc-content/kogito-docs/src/main/asciidoc/configuration/proc_prometheus-metrics-monitoring.adoc:
```diff
@@ -0,0 +1,121 @@
+curl -X GET http://localhost:8080/rest/metrics -H 'content-type: application/json' -H 'accept: application/json'
+----
+
+*<@Cristiano, where might I find an example Prometheus response for Kogito to indlude?>*
```
Ack, thanks!

sterobin commented on doc-content/kogito-docs/src/main/asciidoc/configuration/proc_prometheus-metrics-monitoring.adoc:
```diff
@@ -0,0 +1,121 @@
+
+If the metrics are not available at the defined endpoint, review and verify the {PRODUCT} and Prometheus configurations described in this section.
+
+*<@Cristiano, where might I find an example Prometheus expression browser or Grafana display to use for a screenshot? As shown at the end of the section here: https://access.redhat.com/documentation/en-us/red_hat_process_automation_manager/7.7/html-single/managing_and_monitoring_kie_server/index#prometheus-monitoring-proc_execution-server>*
```
Yes! Exactly what I needed for screenshots and walkthrough. Perfect. Also what I needed in preparation for management console come up. Have bookmarked the video. Thanks Cristiano!

sterobin commented on doc-content/kogito-docs/src/main/asciidoc/configuration/ref_data-index-service-queries.adoc:
```diff
@@ -0,0 +1,420 @@
+----
+--
+
+Correlate data using `metadata` parameter::
```
Right, the double colons format it as you see in the output, with clearer readability in segments. It's a standard formatting we use for look-up items like this, and I'm using it also for consistency with the API look-ups and the like in other docs.

sterobin commented:
@cristianonicolai , implemented your feedback, updated the doc previews, and left one remaining comment above. Good now? Specifically the Prometheus section with those updates.

@tiagodolphine , you good as well? Updated previews linked in description above.

tiagodolphine commented on doc-content/kogito-docs/src/main/asciidoc/configuration/con_jobs-service.adoc:
```diff
@@ -0,0 +1,252 @@
+
+The {PRODUCT} Jobs Service supports Apache Kafka messaging to publish events for each job state transition to a defined Kafka topic. Any application can subscribe to this Kafka topic to receive information about jobs and job state transitions. For example, the {PRODUCT} Data Index Service is subscribed to the Jobs Service Kafka topic so that if you configure and run the Jobs Service, the Data Ind ...
+
+To enable Kafka messaging in the Jobs Service, you specify the property `Dquarkus.profile=events-support` during Jobs Service start-up or add the property to the `src/main/resources/application.properties` file in the Jobs Service project. You must also specify the Kafka bootstrap server with the property `mp.messaging.outgoing.kogito-job-service-job-status-events.bootstrap.servers=__SERVER_ADDRE ...
```
missing "-" on `Dquarkus.profile=events-support` should be `-Dquarkus.profile=events-support`

tiagodolphine replied:
The profile cannot be changed on application.properties, only when starting or with environment variable like `QUARKUS_PROFILE=events-support`

sterobin replied:
Ah, sorry. For some reason I read that to mean the same thing, like when enabling persistence (section before it). For persistence you can add the property at startup or in the app properties file. Thought was like that too.

Fixed as follows:

>To enable Kafka messaging in the Jobs Service, you specify the property `-Dquarkus.profile=events-support` during Jobs Service start-up or add the environment variable `QUARKUS_PROFILE=events-support`.

sterobin commented:
@tiagodolphine , sorry to drag this out, but in finalizing the content, I got a bit turned around with the current form of the Jobs Service chapter and after re-reading some more, I reorganized to highlight Infinispan persistence and Kafka messaging as supplemental procedures, following the core procedure of actually using the Jobs Service. There was just too much mixture of concept vs. procedure how it was.

Can you take a quick last glance at the updated [Kogito Jobs Service](http://file.rdu.redhat.com/~sterobin/BXMSDOC-5392/#con_jobs-service_kogito-configuring) chapter to confirm you're still good with what I've done? **Refresh to get the latest.**

Also, neither Infinispan persistence nor Kafka messaging are required in using the Jobs Service, just optional if the users wishes, correct? Meaning, these two sections do not need to be prerequisites (and this moved around) for the core procedure on "Using the Kogito Jobs Service as a timer service", correct?

tiagodolphine commented:
@sterobin I'd say enabling persistence is not mandatory but is recommended otherwise if jobs-service is restarted all jobs will be lost. Kafka, on the other hand, is optional, and it is more related to expose the data from jobs service, so it is not as essential as persistence IMO.

sterobin commented:
@tiagodolphine , got it. Okay. So I'll leave my updates then and have just added a bit on Inifinspan in the procedure for using the Jobs Service as well, as an aid:

```
. In a command terminal, navigate to the directory location of the downloaded `jobs-service-__VERSION__-runner.jar` file and enter the following command to run the Jobs Service:
+
--
.Running the Jobs Service
[source,subs="+quotes"]
----
$ java -jar jobs-service-__VERSION__-runner.jar
----

To enable Infinispan persistence so that job data is preserved across application restarts, you can specify the following start-up property:

.Enabling Infinispan persistence
[source,subs="+quotes"]
----
$ java  \
    -Dkogito.job-service.persistence=infinispan  \
    -jar jobs-service-__VERSION__-runner.jar
----

To change the logging level of the Jobs Service, such as for debugging, you can specify the following start-up properties:

.Modifying Jobs Service logging level for debugging
[source,subs="+quotes"]
----
$ java  \
  -Dquarkus.log.console.level=DEBUG -Dquarkus.log.category.\"org.kie.kogito\".min-level=DEBUG  \
  -Dquarkus.log.category.\"org.kie.kogito\".level=DEBUG  \
  -jar jobs-service-__VERSION__-runner.jar
----
--
```

Also shown in the updated doc preview links above.

If you have any other feedback, let me know. Otherwise moving this forward. Thank you!

tiagodolphine commented:
> @tiagodolphine , got it. Okay. So I'll leave my updates then but have tweaked the main "Using the Jobs Service" proedure to just tell users to use Infinispan persistence, since doesn't make sense to use the service without it:
> 
> ```
> . In a command terminal, navigate to the directory location of the downloaded `jobs-service-__VERSION__-runner.jar` file and enter the following command to run the Jobs Service with Infinispan persistence enabled:
> +
> --
> .Running the Jobs Service with Infinispan persistence enabled
> [source,subs="+quotes"]
> ----
> $ java  \
>     -Dkogito.job-service.persistence=infinispan  \
>     -jar jobs-service-__VERSION__-runner.jar
> ----
> 
> Infinispan persistence enables the job data to be preserved across application restarts. If you do not use Infinispan persistence, the Jobs Service uses the default in-memory storage and all job information is lost between application restarts.
> 
> To change the logging level of the Jobs Service, such as for debugging, you can specify the following start-up properties:
> 
> .Modifying Jobs Service logging level for debugging
> [source,subs="+quotes"]
> ----
> $ java  \
>   -Dquarkus.log.console.level=DEBUG -Dquarkus.log.category.\"org.kie.kogito\".min-level=DEBUG  \
>   -Dquarkus.log.category.\"org.kie.kogito\".level=DEBUG  \
>   -jar jobs-service-__VERSION__-runner.jar
> ----
> --
> ```
> 
> Also shown in the updated doc preview links above.
> 
> If you have any other feedback, let me know. Otherwise moving this forward. Thank you!

@sterobin looks good to me

sterobin commented on doc-content/kogito-docs/src/main/asciidoc/configuration/con_kogito-runtime-events.adoc:
```diff
@@ -0,0 +1,204 @@
+
+== User task instance events
+
+If an executed request (unit of work) in a process instance interacts with any user tasks, an event is generated for each user task and contains information for each task instance, such as the following information:
```
Yeah, makes sense. Done.

sterobin commented on doc-content/kogito-docs/src/main/asciidoc/configuration/ref_kogito-runtime-properties.adoc:
```diff
@@ -0,0 +1,143 @@
+
+= {PRODUCT} runtime properties quick reference
+
+The following table serves as a quick reference for commonly used runtime configuration properties supported by {PRODUCT}. You can define these properties in the `src/main/resources/application.properties` file of the relevant {PRODUCT} project or using the `-D` prefix during application start-up.
```
Done

sterobin commented on assemblies/assembly_kogito-configuring/main.adoc:
```diff
@@ -0,0 +1,46 @@
+
+[IMPORTANT]
+====
+{PRODUCT} is currently for Developer Preview only. Developer Preview releases contain features that might not be fully tested and are not in a production-ready state. Customers are advised to avoid running the software for production or business-critical workloads. Customers are not able to open support cases against Developer Preview releases. Red Hat does not commit to fixing any reported issue ...
```
Mostly. Red Hat has like bullet points of what Dev Preview means, but not consolidated statement, so this is what I created based on those points. It's being approved and reworded a bit by PM.

sterobin commented on doc-content/kogito-docs/src/main/asciidoc/bpmn/ref_bpmn-tasks.adoc:
```diff
@@ -54,37 +54,37 @@ In the {PRODUCT} BPMN designer, you can modify the following properties for a se
 
 | *On Entry Action*
-| Enter a Java, JavaScript, or MVEL script that specifies an action at the start of the task.
+| Defines a Java, JavaScript, or MVEL script that directs an action at the start of the task.
```
Good point. It's slightly different in that the Script option doesn't differentiate with on entry or exit (start or end of the task), just when the task is executed. I suppose it's equivalent to on entry then. But I don't want to detail that in case that's technically not right or something. For now I've at least made the descriptions a bit more parallel:

>| *Script*
>| Defines a Java, JavaScript, or MVEL script to be executed by the task and specifies the script type.
